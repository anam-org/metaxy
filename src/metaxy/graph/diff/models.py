"""Core data models for graph rendering."""

from enum import Enum
from typing import TYPE_CHECKING, Any

from pydantic import Field
from typing_extensions import Self

from metaxy.models.bases import FrozenBaseModel
from metaxy.models.types import FeatureKey, FieldKey

if TYPE_CHECKING:
    from metaxy.models.feature import FeatureGraph


class NodeStatus(str, Enum):
    """Status of a node in a diff view."""

    NORMAL = "normal"  # Normal node (not in diff mode)
    UNCHANGED = "unchanged"  # Unchanged in diff
    ADDED = "added"  # Added in diff
    REMOVED = "removed"  # Removed in diff
    CHANGED = "changed"  # Changed in diff


class FieldNode(FrozenBaseModel):
    """Represents a field within a feature node.

    Attributes:
        key: Field key
        version: Current field version hash
        old_version: Previous field version hash (for diffs)
        code_version: Code version (if available)
        status: Field status (for diff rendering)
    """

    key: FieldKey
    version: str | None = None  # None if field was removed
    old_version: str | None = None  # For diff mode
    code_version: int | None = None
    status: NodeStatus = NodeStatus.NORMAL


class GraphNode(FrozenBaseModel):
    """Represents a feature node in the graph.

    Attributes:
        key: Feature key
        version: Current feature version hash
        old_version: Previous feature version hash (for diffs)
        code_version: Code version (if available)
        fields: List of field nodes
        dependencies: List of feature keys this node depends on
        status: Node status (for diff rendering)
        project: Project name this feature belongs to
        metadata: Additional custom metadata
    """

    key: FeatureKey
    version: str | None = None  # None if feature was removed
    old_version: str | None = None  # For diff mode
    code_version: int | None = None
    fields: list[FieldNode] = Field(default_factory=list)
    dependencies: list[FeatureKey] = Field(default_factory=list)
    status: NodeStatus = NodeStatus.NORMAL
    project: str | None = None  # Project name (None for legacy features)
    metadata: dict[str, Any] = Field(default_factory=dict)


class EdgeData(FrozenBaseModel):
    """Represents an edge between two nodes.

    Attributes:
        from_key: Source feature key (dependency)
        to_key: Target feature key (dependent)
    """

    from_key: FeatureKey
    to_key: FeatureKey


class GraphData(FrozenBaseModel):
    """Container for complete graph structure.

    This is the unified data model used by all renderers.

    Attributes:
        nodes: Map from feature key string to GraphNode
        edges: List of edges
        snapshot_version: Optional snapshot version
        old_snapshot_version: Optional old snapshot version (for diffs)
    """

    nodes: dict[str, GraphNode]  # Key is feature_key.to_string()
    edges: list[EdgeData] = Field(default_factory=list)
    snapshot_version: str | None = None
    old_snapshot_version: str | None = None  # For diff mode

    def get_node(self, key: FeatureKey) -> GraphNode | None:
        """Get node by feature key.

        Args:
            key: Feature key to lookup

        Returns:
            GraphNode if found, None otherwise
        """
        return self.nodes.get(key.to_string())

    def get_nodes_by_status(self, status: NodeStatus) -> list[GraphNode]:
        """Get all nodes with a specific status.

        Args:
            status: Status to filter by

        Returns:
            List of nodes with matching status
        """
        return [node for node in self.nodes.values() if node.status == status]

    def to_struct(self) -> dict[str, Any]:
        """Serialize to struct (native Python types for storage).

        Note: This uses custom serialization instead of Pydantic's model_dump() because:
        1. Polars struct columns require specific type conversions (e.g., None → "" for strings, None → 0 for ints)
        2. Custom types (FeatureKey, FieldKey) need explicit string conversion for storage
        3. The storage schema is a separate concern from the domain model's Python representation
        4. Different storage backends may need different serialization formats in the future

        Returns:
            Dict with structure compatible with Polars struct type
        """
        nodes_list = []
        for node in self.nodes.values():
            fields_list = []
            for field in node.fields:
                fields_list.append(
                    {
                        "key": field.key.to_string(),
                        "version": field.version if field.version is not None else "",
                        "code_version": field.code_version
                        if field.code_version is not None
                        else 0,
                    }
                )

            nodes_list.append(
                {
                    "key": node.key.to_string(),
                    "version": node.version if node.version is not None else "",
                    "code_version": node.code_version
                    if node.code_version is not None
                    else 0,
                    "fields": fields_list,
                    "dependencies": [dep.to_string() for dep in node.dependencies],
                    "project": node.project if node.project is not None else "",
                }
            )

        edges_list = []
        for edge in self.edges:
            edges_list.append(
                {
                    "from_key": edge.from_key.to_string(),
                    "to_key": edge.to_key.to_string(),
                }
            )

        result: dict[str, Any] = {
            "nodes": nodes_list,
            "edges": edges_list,
        }

        # Include snapshot_version if present
        if self.snapshot_version is not None:
            result["snapshot_version"] = self.snapshot_version

        # Include old_snapshot_version if present (for diffs)
        if self.old_snapshot_version is not None:
            result["old_snapshot_version"] = self.old_snapshot_version

        return result

    @classmethod
    def from_struct(cls, struct_data: dict[str, Any]) -> Self:
        """Deserialize from struct.

        Args:
            struct_data: Dict with structure from to_struct()

        Returns:
            GraphData instance
        """
        nodes = {}
        for node_data in struct_data["nodes"]:
            fields = []
            for field_data in node_data["fields"]:
                fields.append(
                    FieldNode(
                        key=FieldKey(field_data["key"].split("/")),
                        version=field_data["version"]
                        if field_data["version"]
                        else None,
                        code_version=field_data["code_version"]
                        if field_data["code_version"] != 0
                        else None,
                    )
                )

            node = GraphNode(
                key=FeatureKey(node_data["key"].split("/")),
                version=node_data["version"] if node_data["version"] else None,
                code_version=node_data["code_version"]
                if node_data["code_version"] != 0
                else None,
                fields=fields,
                dependencies=[
                    FeatureKey(dep.split("/")) for dep in node_data["dependencies"]
                ],
                project=node_data.get("project") if node_data.get("project") else None,
            )
            nodes[node_data["key"]] = node

        edges = []
        for edge_data in struct_data["edges"]:
            edges.append(
                EdgeData(
                    from_key=FeatureKey(edge_data["from_key"].split("/")),
                    to_key=FeatureKey(edge_data["to_key"].split("/")),
                )
            )

        # Extract snapshot_version if present
        snapshot_version = struct_data.get("snapshot_version")

        # Extract old_snapshot_version if present (for diffs)
        old_snapshot_version = struct_data.get("old_snapshot_version")

        return cls(
            nodes=nodes,
            edges=edges,
            snapshot_version=snapshot_version,
            old_snapshot_version=old_snapshot_version,
        )

    @classmethod
    def from_feature_graph(cls, graph: "FeatureGraph") -> "GraphData":
        """Convert a FeatureGraph to GraphData.

        Args:
            graph: FeatureGraph instance

        Returns:
            GraphData with all nodes and edges
        """
        from metaxy.models.plan import FQFieldKey

        nodes: dict[str, GraphNode] = {}
        edges: list[EdgeData] = []

        # Convert each feature to a GraphNode
        for feature_key, feature_cls in graph.features_by_key.items():
            feature_key_str = feature_key.to_string()
            spec = feature_cls.spec

            # Get feature version
            feature_version = graph.get_feature_version(feature_key)

            # Convert fields
            field_nodes: list[FieldNode] = []
            if spec.fields:
                for field_spec in spec.fields:
                    # Compute field version
                    fq_field_key = FQFieldKey(feature=feature_key, field=field_spec.key)
                    field_version = graph.get_field_version(fq_field_key)

                    field_node = FieldNode(
                        key=field_spec.key,
                        version=field_version,
                        code_version=field_spec.code_version,
                        status=NodeStatus.NORMAL,
                    )
                    field_nodes.append(field_node)

            # Extract dependencies
            dependencies: list[FeatureKey] = []
            if spec.deps:
                dependencies = [dep.key for dep in spec.deps]

            # Get project from feature class
            feature_project = feature_cls.project  # type: ignore[attr-defined]

            # Create node
            node = GraphNode(
                key=feature_key,
                version=feature_version,
                code_version=spec.code_version,
                fields=field_nodes,
                dependencies=dependencies,
                status=NodeStatus.NORMAL,
                project=feature_project,
            )
            nodes[feature_key_str] = node

            # Create edges
            for dep_key in dependencies:
                edges.append(EdgeData(from_key=dep_key, to_key=feature_key))

        return cls(
            nodes=nodes,
            edges=edges,
            snapshot_version=graph.snapshot_version,
        )

    @classmethod
    def from_merged_diff(cls, merged_data: dict[str, Any]) -> "GraphData":
        """Convert merged diff data to GraphData.

        Args:
            merged_data: Merged diff data from GraphDiffer.create_merged_graph_data()

        Returns:
            GraphData with status annotations
        """
        from metaxy.graph.diff.diff_models import FieldChange

        nodes: dict[str, GraphNode] = {}
        edges: list[EdgeData] = []

        # Convert nodes
        for feature_key_str, node_data in merged_data["nodes"].items():
            # Parse feature key
            feature_key = FeatureKey(feature_key_str.split("/"))

            # Map status strings to NodeStatus enum
            status_str = node_data["status"]
            if status_str == "added":
                status = NodeStatus.ADDED
            elif status_str == "removed":
                status = NodeStatus.REMOVED
            elif status_str == "changed":
                status = NodeStatus.CHANGED
            elif status_str == "unchanged":
                status = NodeStatus.UNCHANGED
            else:
                status = NodeStatus.NORMAL

            # Convert fields
            fields_dict = node_data.get("fields", {})
            field_changes_list = node_data.get("field_changes", [])

            # Build field change map for quick lookup
            field_change_map: dict[str, FieldChange] = {}
            for fc in field_changes_list:
                if isinstance(fc, FieldChange):
                    field_change_map[fc.field_key.to_string()] = fc

            # Get all field keys (from both current fields and removed fields in changes)
            all_field_keys = set(fields_dict.keys())
            all_field_keys.update(field_change_map.keys())

            field_nodes: list[FieldNode] = []
            for field_key_str in all_field_keys:
                # Parse field key
                field_key = FieldKey(field_key_str.split("/"))

                # Determine field status and versions
                if field_key_str in field_change_map:
                    fc = field_change_map[field_key_str]
                    if fc.is_added:
                        field_status = NodeStatus.ADDED
                        field_version = fc.new_version
                        old_field_version = None
                    elif fc.is_removed:
                        field_status = NodeStatus.REMOVED
                        field_version = None
                        old_field_version = fc.old_version
                    elif fc.is_changed:
                        field_status = NodeStatus.CHANGED
                        field_version = fc.new_version
                        old_field_version = fc.old_version
                    else:
                        field_status = NodeStatus.UNCHANGED
                        field_version = fc.new_version or fc.old_version
                        old_field_version = None
                else:
                    # Unchanged field
                    field_status = NodeStatus.UNCHANGED
                    field_version = fields_dict.get(field_key_str)
                    old_field_version = None

                field_node = FieldNode(
                    key=field_key,
                    version=field_version,
                    old_version=old_field_version,
                    status=field_status,
                )
                field_nodes.append(field_node)

            # Parse dependencies
            dependencies = [
                FeatureKey(dep_str.split("/"))
                for dep_str in node_data.get("dependencies", [])
            ]

            # Create node
            node = GraphNode(
                key=feature_key,
                version=node_data.get("new_version"),
                old_version=node_data.get("old_version"),
                fields=field_nodes,
                dependencies=dependencies,
                status=status,
            )
            nodes[feature_key_str] = node

        # Convert edges
        for edge_dict in merged_data["edges"]:
            from_key = FeatureKey(edge_dict["from"].split("/"))
            to_key = FeatureKey(edge_dict["to"].split("/"))
            edges.append(EdgeData(from_key=from_key, to_key=to_key))

        return cls(
            nodes=nodes,
            edges=edges,
        )
