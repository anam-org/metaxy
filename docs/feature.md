## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
