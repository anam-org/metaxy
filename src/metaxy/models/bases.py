import pydantic


class FrozenBaseModel(pydantic.BaseModel):
    model_config = pydantic.ConfigDict(frozen=True)
