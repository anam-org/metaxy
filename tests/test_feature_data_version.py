from metaxy.models.feature import Feature, FeatureGraph
from metaxy.models.feature_spec import FeatureDep, FeatureSpec
from metaxy.models.field import FieldDep, FieldSpec, SpecialFieldDep
from metaxy.models.types import FeatureKey, FieldKey


def test_single_feature_data_version(snapshot, graph: FeatureGraph):
    """Test feature with no dependencies."""

    class MyFeature(
        Feature, spec=FeatureSpec(key=FeatureKey(["my_feature"]), deps=None)
    ):
        pass

    assert MyFeature.data_version() == snapshot


def test_feature_with_multiple_fields(snapshot, graph: FeatureGraph):
    """Test feature with multiple fields, each with different code versions."""

    class VideoFeature(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["video"]),
            deps=None,
            fields=[
                FieldSpec(key=FieldKey(["frames"]), code_version="1"),
                FieldSpec(key=FieldKey(["audio"]), code_version="2"),
                FieldSpec(key=FieldKey(["metadata"]), code_version="5"),
            ],
        ),
    ):
        pass

    assert VideoFeature.data_version() == snapshot


def test_linear_dependency_chain(snapshot, graph: FeatureGraph):
    """Test A -> B -> C linear dependency chain."""

    class FeatureA(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["a"]),
            deps=None,
            fields=[FieldSpec(key=FieldKey(["raw"]), code_version="1")],
        ),
    ):
        pass

    class FeatureB(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["b"]),
            deps=[FeatureDep(key=FeatureKey(["a"]))],
            fields=[FieldSpec(key=FieldKey(["processed"]), code_version="2")],
        ),
    ):
        pass

    class FeatureC(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["c"]),
            deps=[FeatureDep(key=FeatureKey(["b"]))],
            fields=[FieldSpec(key=FieldKey(["final"]), code_version="3")],
        ),
    ):
        pass

    versions = {
        "a": FeatureA.data_version(),
        "b": FeatureB.data_version(),
        "c": FeatureC.data_version(),
    }

    assert versions == snapshot


def test_diamond_dependency_graph(snapshot, graph: FeatureGraph):
    """Test diamond dependency: A -> B, A -> C, B -> D, C -> D."""

    class Root(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["root"]),
            deps=None,
            fields=[FieldSpec(key=FieldKey(["data"]), code_version="1")],
        ),
    ):
        pass

    class BranchLeft(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["branch_left"]),
            deps=[FeatureDep(key=FeatureKey(["root"]))],
            fields=[FieldSpec(key=FieldKey(["left_processed"]), code_version="2")],
        ),
    ):
        pass

    class BranchRight(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["branch_right"]),
            deps=[FeatureDep(key=FeatureKey(["root"]))],
            fields=[FieldSpec(key=FieldKey(["right_processed"]), code_version="3")],
        ),
    ):
        pass

    class Merged(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["merged"]),
            deps=[
                FeatureDep(key=FeatureKey(["branch_left"])),
                FeatureDep(key=FeatureKey(["branch_right"])),
            ],
            fields=[FieldSpec(key=FieldKey(["fusion"]), code_version="4")],
        ),
    ):
        pass

    versions = {
        "root": Root.data_version(),
        "branch_left": BranchLeft.data_version(),
        "branch_right": BranchRight.data_version(),
        "merged": Merged.data_version(),
    }

    assert versions == snapshot


def test_specific_field_dependencies(snapshot, graph: FeatureGraph):
    """Test feature with specific field-level dependencies."""

    class MultiField(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["multi"]),
            deps=None,
            fields=[
                FieldSpec(key=FieldKey(["frames"]), code_version="1"),
                FieldSpec(key=FieldKey(["audio"]), code_version="2"),
                FieldSpec(key=FieldKey(["text"]), code_version="3"),
            ],
        ),
    ):
        pass

    class Selective(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["selective"]),
            deps=[FeatureDep(key=FeatureKey(["multi"]))],
            fields=[
                # Only depends on frames
                FieldSpec(
                    key=FieldKey(["visual"]),
                    code_version="10",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["multi"]),
                            fields=[FieldKey(["frames"])],
                        )
                    ],
                ),
                # Only depends on audio
                FieldSpec(
                    key=FieldKey(["audio_only"]),
                    code_version="11",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["multi"]),
                            fields=[FieldKey(["audio"])],
                        )
                    ],
                ),
                # Depends on frames and text
                FieldSpec(
                    key=FieldKey(["mixed"]),
                    code_version="12",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["multi"]),
                            fields=[
                                FieldKey(["frames"]),
                                FieldKey(["text"]),
                            ],
                        )
                    ],
                ),
            ],
        ),
    ):
        pass

    versions = {
        "multi": MultiField.data_version(),
        "selective": Selective.data_version(),
    }

    assert versions == snapshot


def test_complex_multi_level_graph(snapshot, graph: FeatureGraph):
    """Test complex graph with multiple levels and mixed dependency types."""

    # Level 0: Two root features
    class RawVideoData(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["raw_video"]),
            deps=None,
            fields=[
                FieldSpec(key=FieldKey(["frames"]), code_version="1"),
                FieldSpec(key=FieldKey(["audio"]), code_version="1"),
            ],
        ),
    ):
        pass

    class RawMetadata(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["raw_metadata"]),
            deps=None,
            fields=[FieldSpec(key=FieldKey(["info"]), code_version="1")],
        ),
    ):
        pass

    # Level 1: Process video
    class ProcessedVideo(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["processed_video"]),
            deps=[FeatureDep(key=FeatureKey(["raw_video"]))],
            fields=[
                FieldSpec(
                    key=FieldKey(["enhanced_frames"]),
                    code_version="5",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["raw_video"]),
                            fields=[FieldKey(["frames"])],
                        )
                    ],
                ),
                FieldSpec(
                    key=FieldKey(["normalized_audio"]),
                    code_version="3",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["raw_video"]),
                            fields=[FieldKey(["audio"])],
                        )
                    ],
                ),
            ],
        ),
    ):
        pass

    # Level 2: Analysis combining multiple sources
    class Analysis(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["analysis"]),
            deps=[
                FeatureDep(key=FeatureKey(["processed_video"])),
                FeatureDep(key=FeatureKey(["raw_metadata"])),
            ],
            fields=[
                # Depends on all upstream
                FieldSpec(
                    key=FieldKey(["full_analysis"]),
                    code_version="10",
                    deps=SpecialFieldDep.ALL,
                ),
                # Depends only on enhanced frames and metadata
                FieldSpec(
                    key=FieldKey(["visual_metadata"]),
                    code_version="7",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["processed_video"]),
                            fields=[FieldKey(["enhanced_frames"])],
                        ),
                        FieldDep(
                            feature_key=FeatureKey(["raw_metadata"]),
                            fields=SpecialFieldDep.ALL,
                        ),
                    ],
                ),
            ],
        ),
    ):
        pass

    versions = {
        "raw_video": RawVideoData.data_version(),
        "raw_metadata": RawMetadata.data_version(),
        "processed_video": ProcessedVideo.data_version(),
        "analysis": Analysis.data_version(),
    }

    assert versions == snapshot


def test_code_version_changes_propagate(snapshot, graph: FeatureGraph):
    """Test that changing code version in upstream affects downstream."""

    class Base(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["base"]),
            deps=None,
            fields=[FieldSpec(key=FieldKey(["data"]), code_version="100")],
        ),
    ):
        pass

    class Derived(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["derived"]),
            deps=[FeatureDep(key=FeatureKey(["base"]))],
            fields=[FieldSpec(key=FieldKey(["processed"]), code_version="1")],
        ),
    ):
        pass

    versions = {
        "base": Base.data_version(),
        "derived": Derived.data_version(),
    }

    assert versions == snapshot


def test_multiple_fields_different_deps(snapshot, graph: FeatureGraph):
    """Test feature where different fields have completely different dependency sets."""

    class FeatureX(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["x"]),
            deps=None,
            fields=[
                FieldSpec(key=FieldKey(["x1"]), code_version="1"),
                FieldSpec(key=FieldKey(["x2"]), code_version="2"),
            ],
        ),
    ):
        pass

    class FeatureY(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["y"]),
            deps=None,
            fields=[
                FieldSpec(key=FieldKey(["y1"]), code_version="3"),
                FieldSpec(key=FieldKey(["y2"]), code_version="4"),
            ],
        ),
    ):
        pass

    class FeatureZ(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["z"]),
            deps=[
                FeatureDep(key=FeatureKey(["x"])),
                FeatureDep(key=FeatureKey(["y"])),
            ],
            fields=[
                # First field only depends on FeatureX
                FieldSpec(
                    key=FieldKey(["z1"]),
                    code_version="20",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["x"]),
                            fields=SpecialFieldDep.ALL,
                        )
                    ],
                ),
                # Second field only depends on FeatureY
                FieldSpec(
                    key=FieldKey(["z2"]),
                    code_version="21",
                    deps=[
                        FieldDep(
                            feature_key=FeatureKey(["y"]),
                            fields=SpecialFieldDep.ALL,
                        )
                    ],
                ),
                # Third field depends on both
                FieldSpec(
                    key=FieldKey(["z3"]),
                    code_version="22",
                    deps=SpecialFieldDep.ALL,
                ),
            ],
        ),
    ):
        pass

    versions = {
        "x": FeatureX.data_version(),
        "y": FeatureY.data_version(),
        "z": FeatureZ.data_version(),
    }

    # Verify that different fields have different versions
    z_versions = versions["z"]
    assert len(set(z_versions.values())) == 3, (
        "All fields should have different versions"
    )

    assert versions == snapshot
