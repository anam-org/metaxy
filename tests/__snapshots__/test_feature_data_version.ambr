# serializer version: 1
# name: test_code_version_changes_propagate
  dict({
    'base': dict({
      'data': '6dd71299f6e93f91b62ae9a93ec1a030b83d92c5ab6917ca3643c5889e796a72',
    }),
    'derived': dict({
      'processed': '8fcd79f1e165b9b5b0b72f9a63103808ea9e96439dfdd37791c2b303c84b5e67',
    }),
  })
# ---
# name: test_complex_multi_level_graph
  dict({
    'analysis': dict({
      'full_analysis': 'f11c09bafd771ac31e4e9984d0ce1a99f589d16adfe5565303de784b1a149775',
      'visual_metadata': '8527b2ddca8b3b8cd73a760f5841e9afda3670cb12b42f5e66758317c8cc72e3',
    }),
    'processed_video': dict({
      'enhanced_frames': 'bce72e4f538d9ff9e6934ffc3ae43b792face388fba1f828e08dc84633f2ca38',
      'normalized_audio': 'ffccf798b5f6dee659197570d5d9231eed1cb9ccf5bb79bf4c6cf00529e7c0d2',
    }),
    'raw_metadata': dict({
      'info': '3e146e4d524d618ea73b6b0232fd05e462e3224ba1aaf9f4e564efd2cab5d4b2',
    }),
    'raw_video': dict({
      'audio': '678d3e83d66dd32559d518c9d70f1101be85cfd2a3d4a2e8cd16bb49427cc336',
      'frames': 'be12e9e4f2878176fa759e9505eae710188527e656b6786af444bcf1e979df1d',
    }),
  })
# ---
# name: test_diamond_dependency_graph
  dict({
    'branch_left': dict({
      'left_processed': '8499ba90006b890f1e153ba9f022dac81da2183165b8edcb51899cb7c866d38b',
    }),
    'branch_right': dict({
      'right_processed': '4cd36bc6c21ffd090cacb4d0231952c330243216cc50805307f926a32e4d216d',
    }),
    'merged': dict({
      'fusion': 'ee1f6dd3d1a96c98bb3ea8bec3854e47164487a4525b0b04b0c993f32d4eb592',
    }),
    'root': dict({
      'data': '636edb1af7d998b6976f6e11595357bb75a5a596661994ca8145955f850e2cf6',
    }),
  })
# ---
# name: test_feature_with_multiple_fields
  dict({
    'audio': 'ab11f4a65d39f51402bc4eb224822d38ddcb24831ac53a5b124c333a42a5f4a3',
    'frames': 'd6e150f0fd72262c17770d4886290fa8ca25b5d4516389a64a654db718b0a8dd',
    'metadata': '1140d01855d890bb4dfa521c9fb1114a5a861a3a6556ffa81cc7d5a28e1d3c9d',
  })
# ---
# name: test_linear_dependency_chain
  dict({
    'a': dict({
      'raw': '1075ae9f96a59488bacfc0a20f799d232243ea26a07079f4227ee3299c90f597',
    }),
    'b': dict({
      'processed': '936881be29e23b022f89486a25bd67b34abe2bc10c6a6d46339c724882918fee',
    }),
    'c': dict({
      'final': '795f3135ad8a996a2d9de4398d710befd48c73ba3a66ced0751e0cf82bb56cdc',
    }),
  })
# ---
# name: test_multiple_fields_different_deps
  dict({
    'x': dict({
      'x1': 'f040ee10ef63d16928f823dcaddccbb837488070747b37152a279430e9ff2300',
      'x2': '714066d27ea09f099fe2bb56b01ba7432cc579d8b08f30cac017dce0802003ef',
    }),
    'y': dict({
      'y1': '1af01721d81c3e15bbff016e9e91bcfba968ba3a420c45daf3c2824c2f432b81',
      'y2': '9a5b36fb708e70c11c6d4759a19551d14eb87df6c86b7af410a259b741e276ed',
    }),
    'z': dict({
      'z1': '419a88d50544185f047adbaeedd448718b9a7df4481f8bf94dcb33753b24fd48',
      'z2': '954f9936ceccd687fe396d4c871d464b237a965d786934cb9354c9461696dbb0',
      'z3': '2de311f286c82212d019c42aaff6d08e7c4282aec2a6521a18de8befccaef11d',
    }),
  })
# ---
# name: test_single_feature_data_version
  dict({
    'default': '1428690189d4145891f390b4af374c266863c5d3a0a6b24e81d73edf399cc33f',
  })
# ---
# name: test_specific_field_dependencies
  dict({
    'multi': dict({
      'audio': 'e8586a193e228d0876163543a10104afa31026e7b38955ca7ef3c2045fc019f1',
      'frames': 'e32699cc24a597db2388d427f1cd5e60d176994f7ceefaa697b65f1ea0fc194c',
      'text': 'f067c22d1664e3b315388100e3bb10d9f8feab9df3a011517545c7fe1db3fb19',
    }),
    'selective': dict({
      'audio_only': '32a990ed6c83735144f87b2cc135b55068e6171fc0d554458b1e6b5e41ced7a0',
      'mixed': '9dfac4f2835e9dd026689e3771db6bc22dca5208a24dc9b5dc8443b5ed56e0e1',
      'visual': 'f9f46bc326619513e760b2ccaf5c2834e889aa963ad380f1d5b6ee0694c52275',
    }),
  })
# ---
